### PR TITLE
protocol/bc: return errors when writing transactions

### PR DESCRIPTION
**Pull request by tessr**
Part of #349. Wraps up some todos.

### CODE REVIEW

File: protocol/bc/hash.go [hunk 1]
```diff
@@ -96,12 +96,13 @@ func ParseHash(s string) (h Hash, err error) {
 	return h, errors.Wrap(err, "decode hex")
 }
 
-func writeFastHash(w io.Writer, d []byte) {
+func WriteFastHash(w io.Writer, d []byte) error {
```
Comment (erykwalder) [inline protocol/bc/hash.go]: does this need to be exported?

File: protocol/bc/transaction.go [hunk 2]
```diff
@@ -271,39 +271,67 @@ func (tx *TxData) MarshalText() ([]byte, error) {
...
 	ew := errors.NewWriter(w)
-	tx.writeTo(ew, serRequired)
+	err := tx.writeTo(ew, serRequired)
+	if err != nil {
+		return ew.Written(), ew.Err()
+	}
 	return ew.Written(), ew.Err()
 }
 
 // assumes w has sticky errors
```
Comment (erykwalder) [inline protocol/bc/transaction.go]: all comments like this are no longer valid/necessary

File: protocol/bc/transaction.go [hunk 3]
```diff
@@ -271,39 +271,67 @@ func (tx *TxData) MarshalText() ([]byte, error) {
...
+		err = ti.writeTo(w, serflags)
+		if err != nil {
+			return errors.Wrapf(err, "writing tx input %d", i)
+		}
 	}
 
-	blockchain.WriteVarint31(w, uint64(len(tx.Outputs))) // TODO(bobg): check and return error
-	for _, to := range tx.Outputs {
-		to.writeTo(w, serflags)
+	blockchain.WriteVarint31(w, uint64(len(tx.Outputs)))
```
Comment (erykwalder) [inline protocol/bc/transaction.go]: Need to capture `err` here

File: protocol/bc/hash.go (hunk 1 above)
Comment (tessr) [inline protocol/bc/hash.go]: Oh no, I don't think so

File: protocol/bc/transaction.go (hunk 2 above)
Comment (tessr) [inline protocol/bc/transaction.go]: thanks 👍

File: protocol/bc/transaction.go [hunk 4]
```diff
@@ -323,10 +350,10 @@ func (p *Outpoint) WriteTo(w io.Writer) (int64, error) {
 }
 
 // assumes w has sticky errors
```
Comment (erykwalder) [inline protocol/bc/transaction.go]: this assumption is no longer the case

File: protocol/bc/txinput.go [hunk 5]
```diff
@@ -199,13 +199,30 @@ func (t *TxInput) readFrom(r io.Reader, txVersion uint64) (err error) {
 }
 
 // assumes w has sticky errors
```
Comment (erykwalder) [inline protocol/bc/txinput.go]: same here

File: protocol/bc/txoutput.go [hunk 6]
```diff
@@ -55,11 +55,28 @@ func (to *TxOutput) readFrom(r io.Reader, txVersion uint64) (err error) {
 }
 
 // assumes r has sticky errors
```
Comment (erykwalder) [inline protocol/bc/txoutput.go]: same here

**Comment by tessr** [comment]
the formerly-sticky errors's comments have been un-stuck

File: protocol/bc/block.go [hunk 7]
```diff
@@ -106,7 +106,6 @@ func (b *Block) WriteTo(w io.Writer) (int64, error) {
 	return ew.Written(), ew.Err()
 }
 
-// assumes w has sticky errors
```
Comment (erykwalder) [inline protocol/bc/block.go]: This one doesn't check errors, so would assume it has sticky errors. We should either leave comment or check errors.

Comment (tessr) [inline protocol/bc/block.go]: I got a bit overzealous 😐  fixed

Comment (tessr) [inline protocol/bc/block.go]: well, fixed the comment

Comment (tessr) [inline protocol/bc/block.go]: (#349 doesn't touch this, so I for now I will leave this alone too)

**Comment by erykwalder** [comment]
LGTM and sorry for all the back and forth

**Comment by tessr** [comment]
Haha, no worries. Sorry for being sloppy with comments.

**Pull request merged**